packages/components/src/components/SquigglePlayground/index.tsx

```diff
@@ -55,131 +40,72 @@ export const PlaygroundContext = React.createContext<PlaygroundContextShape>({
 
-  const leftPanelBody =
-    selectedTab === "CODE" ? (
-      <div data-testid="squiggle-editor">
-        <CodeEditor
-          ref={editorRef}
-          defaultValue={defaultCode}
-          errors={errors}
-          project={project}
-          sourceId={sourceId}
-          showGutter={true}
-          onChange={setCode}
-          onViewValuePath={(ast) => viewerRef.current?.viewValuePath(ast)}
-          onSubmit={runnerState.run}
-        />
-      </div>
-    ) : selectedTab === "SETTINGS" ? (
-      <div className="px-2 space-y-6">
-        <div className="px-2 py-2">
-          <div className="pb-4">
-            <Button onClick={() => setSelectedTab("CODE")}>Back</Button>
-          </div>
-          <FormProvider {...form}>
-            <PlaygroundSettingsForm />
-          </FormProvider>
-        </div>
-      </div>
-    ) : null;
-
-  const leftPanelRef = useRef<HTMLDivElement | null>(null);
-
-  const leftPanelHeader = (
-    <div className="flex justify-between h-8 bg-slate-50 border-b border-slate-200 overflow-hidden mb-1 px-4">
-      <div className="flex">
-        <RunMenuItem {...runnerState} isRunning={isRunning} />
-        <AutorunnerMenuItem {...runnerState} />
-        <MenuItem
-          onClick={() =>
-            setSelectedTab(selectedTab === "SETTINGS" ? "CODE" : "SETTINGS")
-          }
-          icon={AdjustmentsVerticalIcon}
-          tooltipText="Configuration"
-        />
-        <MenuItem
-          tooltipText={`Format Code (${altKey()}+Shift+f)`}
-          icon={Bars3CenterLeftIcon}
-          onClick={editorRef.current?.format}
-        />
-      </div>
-      <div className="flex items-center">{renderExtraControls?.()}</div>
-    </div>
-  );
+  const leftPanelRef = useRef<LeftPlaygroundPanelHandle>(null);
 
   const getLeftPanelElement = useCallback(
-    () => leftPanelRef.current ?? undefined,
+    () => leftPanelRef.current?.getLeftPanelElement() ?? undefined,
     []
   );
 
   const renderLeft = () => (
-    <div className="h-full flex flex-col" ref={leftPanelRef}>
-      {leftPanelHeader}
-      <div className="flex-1 grid place-content-stretch overflow-auto">
-        {leftPanelBody}
-      </div>
-    </div>
+    <LeftPlaygroundPanel
+      defaultCode={defaultCode}
+      onCodeChange={onCodeChange}
+      settings={settings}
+      onSettingsChange={handleSettingsChange}
+      onOutputChange={setOutput}
+      renderExtraControls={renderExtraControls}
+      renderExtraModal={renderExtraModal}
+      onViewValuePath={(path) => viewerRef.current?.viewValuePath(path)}
+      ref={leftPanelRef}
+    />
   );
 
   const renderRight = () => (
     <DynamicSquiggleViewer
-      squiggleOutput={squiggleOutput}
-      isRunning={isRunning}
-      editor={editorRef.current ?? undefined}
+      squiggleOutput={output.output}
+      isRunning={output.isRunning}
+      // FIXME - this will cause viewer to be rendered twice on initial render
```

Comment:
I'm assuming you don't mean to fix this for this PR

Reply:
Yes. It's a bit hard to fix, and I'm not even sure that the comment is completely correct (what exactly causes the rerender?), more like a note to investigate in the future.